core/lib/zksync_core/src/lib.rs

```diff
@@ -617,6 +625,19 @@ pub async fn initialize_components(
     .await
     .context("add_witness_generator_to_task_futures()")?;
 
+    if components.contains(&Component::BasicWitnessInputProducer) {
+        add_basic_witness_input_producer(
+            &mut task_futures,
+            &connection_pool,
```

Comment:
> We'll need at least 2 connections. 

Aren't they used sequentially (e.g. never at the same time)? It's not a unique connection, but rather a pool with a single connection. Could it be that you just need to `drop` a connection between instantiating the `PostgresStorage`?

Reply:
This is why I mentioned `drop`. If you will `drop` the `PostgresStorage` (or `VmInstance`), it'll drop the connection and it will return to the pool.
One connection still should suffice.

Reply:
You can fetch all this information before instantiating the VM, as suggested in [this comment chain](https://github.com/matter-labs/zksync-era/pull/156#discussion_r1363393559).

Reply:
I mean, the suboptimal code in one place should not be considered a justification to write suboptimal code elsewhere.
Even given your example, if the linked ticket will be fixed, there will be no more justification to use more connections than required in this new component 😅